### ISSUE TITLE
telegraf container prevents docker from removing monitored containers

### ISSUE THREAD

**Issue by m4ce** [influxdata/telegraf#2250]
When telegraf runs under docker and / is mounted as e.g. /hostfs, it suffers from the same issues as cAdvisor (https://github.com/google/cadvisor/issues/771).

When removing a container, the error typically looks like this:

```
Error: Unable to remove filesystem for
74bef250361c7817bee19349c93139621b272bc8f654ae112dd4eb9652af9515:
remove /var/lib/docker/containers/74bef250361c7817bee19349c93139621b272bc8f654ae112dd4eb9652af9515/shm:
Device or resource busy
```

The issue is also described at https://docs.docker.com/engine/admin/troubleshooting_volume_errors

It seems that there's currently no proper fix for the issue.

I've looked into it and adding this the entrypoint.sh should work this around for now:

```shell
if [ -n "$HOST_MOUNT_PREFIX" ]; then
  # this solves https://docs.docker.com/engine/admin/troubleshooting_volume_errors
  mount | awk -v pattern="^(shm|overlay) on $HOST_MOUNT_PREFIX" '{if ($0 ~ pattern) { print $3 }}' | sort -u | while read dir; do
    umount $dir
  done
fi

exec "$@"
```

There are also other issues where I've added some fixes/workarounds around the current issues in running telegraf under docker. It might be worth to collect them all and put them in a README or FAQ page?

**Comment by phemmer** [comment]
> The issue is also described at https://docs.docker.com/engine/admin/troubleshooting_volume_errors

> It seems that there's currently no proper fix for the issue.

If the issue is what you linked, then that means telegraf has open file descriptors on the mount. Telegraf should not be keeping file descriptors open like that. Thus if that is indeed the problem, the proper fix is for telegraf to close its file descriptors.

Can you confirm this is the issue? `lsof $pid_of_telegraf`?

**Comment by m4ce** [comment]
@phemmer, the problem is not really telegraf itself, it's rather the container in which is running that has those volumes mounted. This only happens when running it with -v /:/hostfs:ro or /var/lib/docker:/var/lib/docker:ro for instance. The issue only impacts containers started before telegraf.

**Comment by phemmer** [comment]
Bind mounting `/var/lib/docker` won't hold volumes within it open. This is why the article you link specifically mentions: 

> The problem occurs if the container which bind-mounts /var/lib/docker/ uses statfs or fstatfs on filesystem handles within /var/lib/docker/ and does not close them

Thus something has to be holding open file descriptors.

**Comment by m4ce** [comment]
So, this is a quick test:

Start the test container
```
# docker run -d --entrypoint /bin/sh --name test -it telegraf:1.1.2
82ddbc9328334a5176e7b1d66d77c641b7bd9f95d957e10e70e0b5afdfff186e
```

Start the container which will run telegraf:
```
# docker run -d --privileged --entrypoint /bin/sh -v /:/hostfs:ro --name telegraf -it telegraf:1.1.2
b38acc94c0482a56bb1ff22f1b931d6974835f52e822963bbd3c49bc60741c75
```

Enter the telegraf container:
```
# docker exec -it telegraf /bin/sh
/ # ps ux
PID   USER     TIME   COMMAND
    1 root       0:00 /bin/sh
    7 root       0:00 /bin/sh
   12 root       0:00 ps ux
/ # mount | grep '^overlay' | grep hostfs
overlay on /hostfs/var/lib/docker/overlay/9837d68e2894fe1820e91afc691a5b1141314e5ac88a9b0729c924c70cbb6919/merged type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay/5dcbf254ca4931910ecaab1643f96ab924870482910c525a2c248a528197d858/root,upperdir=/var/lib/docker/overlay/9837d68e2894fe1820e91afc691a5b1141314e5ac88a9b0729c924c70cbb6919/upper,workdir=/var/lib/docker/overlay/9837d68e2894fe1820e91afc691a5b1141314e5ac88a9b0729c924c70cbb6919/work)
overlay on /hostfs/var/lib/docker/overlay/6a5c4cacd33317438dd27c7d0587f6079e2cc0d08445934b2034a8ca1438a75d/merged type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay/8d512f64fac080fd2beb9afc45694ce3dcb9ea0dcd2fec70abee42520dd24dd9/root,upperdir=/var/lib/docker/overlay/6a5c4cacd33317438dd27c7d0587f6079e2cc0d08445934b2034a8ca1438a75d/upper,workdir=/var/lib/docker/overlay/6a5c4cacd33317438dd27c7d0587f6079e2cc0d08445934b2034a8ca1438a75d/work)
overlay on /hostfs/var/lib/docker/overlay/24f8ceafde15b17d1a00faccc64da6821aa1151e36d865efc29300b652624539/merged type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay/7a18a5b0093cd8d8cfb939b4efcf6247a4f4430be3b184bd9ddd2ba232ed25ef/root,upperdir=/var/lib/docker/overlay/24f8ceafde15b17d1a00faccc64da6821aa1151e36d865efc29300b652624539/upper,workdir=/var/lib/docker/overlay/24f8ceafde15b17d1a00faccc64da6821aa1151e36d865efc29300b652624539/work)
overlay on /hostfs/var/lib/docker/overlay/d85f5b3602cf4c328c2edacd067affa6530dafa797afff7e0dd415d3884bf5d2/merged type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay/cf18434fc82007de64cfc2742b6a84a880215bdd3c30996dea8fc3b7b0609ba7/root,upperdir=/var/lib/docker/overlay/d85f5b3602cf4c328c2edacd067affa6530dafa797afff7e0dd415d3884bf5d2/upper,workdir=/var/lib/docker/overlay/d85f5b3602cf4c328c2edacd067affa6530dafa797afff7e0dd415d3884bf5d2/work)
overlay on /hostfs/var/lib/docker/overlay/d3fc3b8df9a3f48193328e4cddada74b0e78ace217d9cde6ccd02143d3ff0c47/merged type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay/dc59567af60f8c70b35c02692e2253904b22c6c9986b3df6fcf95eda090e770c/root,upperdir=/var/lib/docker/overlay/d3fc3b8df9a3f48193328e4cddada74b0e78ace217d9cde6ccd02143d3ff0c47/upper,workdir=/var/lib/docker/overlay/d3fc3b8df9a3f48193328e4cddada74b0e78ace217d9cde6ccd02143d3ff0c47/work)
overlay on /hostfs/var/lib/docker/overlay/2a0f01d5a7455488043920ae09958a59cfc2cb188598e794b79452dee9207f19/merged type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay/dc59567af60f8c70b35c02692e2253904b22c6c9986b3df6fcf95eda090e770c/root,upperdir=/var/lib/docker/overlay/2a0f01d5a7455488043920ae09958a59cfc2cb188598e794b79452dee9207f19/upper,workdir=/var/lib/docker/overlay/2a0f01d5a7455488043920ae09958a59cfc2cb188598e794b79452dee9207f19/work)
overlay on /hostfs/var/lib/docker/overlay/f329e24fff6b080d6810ddd94fc48b90443c30db8568d3bf3c5aedc17ae83d88/merged type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay/dc59567af60f8c70b35c02692e2253904b22c6c9986b3df6fcf95eda090e770c/root,upperdir=/var/lib/docker/overlay/f329e24fff6b080d6810ddd94fc48b90443c30db8568d3bf3c5aedc17ae83d88/upper,workdir=/var/lib/docker/overlay/f329e24fff6b080d6810ddd94fc48b90443c30db8568d3bf3c5aedc17ae83d88/work)
overlay on /hostfs/var/lib/docker/overlay/f329e24fff6b080d6810ddd94fc48b90443c30db8568d3bf3c5aedc17ae83d88/merged type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay/dc59567af60f8c70b35c02692e2253904b22c6c9986b3df6fcf95eda090e770c/root,upperdir=/var/lib/docker/overlay/f329e24fff6b080d6810ddd94fc48b90443c30db8568d3bf3c5aedc17ae83d88/upper,workdir=/var/lib/docker/overlay/f329e24fff6b080d6810ddd94fc48b90443c30db8568d3bf3c5aedc17ae83d88/work)
/ # mount | grep '^shm' | grep hostfs
shm on /hostfs/var/lib/docker/containers/1da0daaaa6a6f134a3f76dd02d9b267ebb954ee7484ec4654ff538bcb1661349/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,size=65536k)
shm on /hostfs/var/lib/docker/containers/b8d2ba560d2ce116e63f2ef5a57cd92c9d25c635c10ce215fec445559e3aab33/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,size=65536k)
shm on /hostfs/var/lib/docker/containers/b73a5f462f043c75c0490f8bbe89a850a48d46b70faf4a385e647878f8780530/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,size=65536k)
shm on /hostfs/var/lib/docker/containers/6c836415563c00614246429ed3157355978561891605a07eda91fcbcef5ddd7c/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,size=65536k)
shm on /hostfs/var/lib/docker/containers/47e387b0d2afcd5c4bb7f4244449082013c93142c48f0f5f42aa8c8a6596362f/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,size=65536k)
shm on /hostfs/var/lib/docker/containers/82ddbc9328334a5176e7b1d66d77c641b7bd9f95d957e10e70e0b5afdfff186e/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,size=65536k)
shm on /hostfs/var/lib/docker/containers/b38acc94c0482a56bb1ff22f1b931d6974835f52e822963bbd3c49bc60741c75/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,size=65536k)
```

Now stop the test container and remove it.
```
# docker stop test
test
# docker rm test
Error response from daemon: Unable to remove filesystem for 82ddbc9328334a5176e7b1d66d77c641b7bd9f95d957e10e70e0b5afdfff186e: remove /var/lib/docker/containers/82ddbc9328334a5176e7b1d66d77c641b7bd9f95d957e10e70e0b5afdfff186e/shm: device or resource busy
```

Enter the telegraf container and unmount it:
```
# docker exec -it telegraf /bin/sh
/ # umount /hostfs/var/lib/docker/containers/82ddbc9328334a5176e7b1d66d77c641b7bd9f95d957e10e70e0b5afdfff186e/shm
#
```

Now try to remove the container again:
```
# docker rm test
Error response from daemon: Driver overlay failed to remove root filesystem 82ddbc9328334a5176e7b1d66d77c641b7bd9f95d957e10e70e0b5afdfff186e: remove /var/lib/docker/overlay/2a0f01d5a7455488043920ae09958a59cfc2cb188598e794b79452dee9207f19/merged: device or resource busy
```

Enter the telegraf container and unmount it:

```
# docker exec -it telegraf /bin/sh
/ # umount /hostfs/var/lib/docker/overlay/2a0f01d5a7455488043920ae09958a59cfc2cb188598e794b79452dee9207f19/merged
```

Try to remove the test container again, should work now:

```
# docker rm test
test
```

Again, this only happens with containers that have started before telegraf.

**Comment by phemmer** [comment]
```
/ # mount | grep '^overlay' | grep hostfs
overlay on /hostfs/var/lib/docker/overlay/9837d68e2894fe1820e91afc691a5b1141314e5ac88a9b0729c924c70cbb6919/merged type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay/5dcbf254ca4931910ecaab1643f96ab924870482910c525a2c248a528197d858/root,upperdir=/var/lib/docker/overlay/9837d68e2894fe1820e91afc691a5b1141314e5ac88a9b0729c924c70cbb6919/upper,workdir=/var/lib/docker/overlay/9837d68e2894fe1820e91afc691a5b1141314e5ac88a9b0729c924c70cbb6919/work)
overlay on /hostfs/var/lib/docker/overlay/6a5c4cacd33317438dd27c7d0587f6079e2cc0d08445934b2034a8ca1438a75d/merged type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay/8d512f64fac080fd2beb9afc45694ce3dcb9ea0dcd2fec70abee42520dd24dd9/root,upperdir=/var/lib/docker/overlay/6a5c4cacd33317438dd27c7d0587f6079e2cc0d08445934b2034a8ca1438a75d/upper,workdir=/var/lib/docker/overlay/6a5c4cacd33317438dd27c7d0587f6079e2cc0d08445934b2034a8ca1438a75d/work)
overlay on /hostfs/var/lib/docker/overlay/24f8ceafde15b17d1a00faccc64da6821aa1151e36d865efc29300b652624539/merged type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay/7a18a5b0093cd8d8cfb939b4efcf6247a4f4430be3b184bd9ddd2ba232ed25ef/root,upperdir=/var/lib/docker/overlay/24f8ceafde15b17d1a00faccc64da6821aa1151e36d865efc29300b652624539/upper,workdir=/var/lib/docker/overlay/24f8ceafde15b17d1a00faccc64da6821aa1151e36d865efc29300b652624539/work)
overlay on /hostfs/var/lib/docker/overlay/d85f5b3602cf4c328c2edacd067affa6530dafa797afff7e0dd415d3884bf5d2/merged type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay/cf18434fc82007de64cfc2742b6a84a880215bdd3c30996dea8fc3b7b0609ba7/root,upperdir=/var/lib/docker/overlay/d85f5b3602cf4c328c2edacd067affa6530dafa797afff7e0dd415d3884bf5d2/upper,workdir=/var/lib/docker/overlay/d85f5b3602cf4c328c2edacd067affa6530dafa797afff7e0dd415d3884bf5d2/work)
overlay on /hostfs/var/lib/docker/overlay/d3fc3b8df9a3f48193328e4cddada74b0e78ace217d9cde6ccd02143d3ff0c47/merged type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay/dc59567af60f8c70b35c02692e2253904b22c6c9986b3df6fcf95eda090e770c/root,upperdir=/var/lib/docker/overlay/d3fc3b8df9a3f48193328e4cddada74b0e78ace217d9cde6ccd02143d3ff0c47/upper,workdir=/var/lib/docker/overlay/d3fc3b8df9a3f48193328e4cddada74b0e78ace217d9cde6ccd02143d3ff0c47/work)
overlay on /hostfs/var/lib/docker/overlay/2a0f01d5a7455488043920ae09958a59cfc2cb188598e794b79452dee9207f19/merged type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay/dc59567af60f8c70b35c02692e2253904b22c6c9986b3df6fcf95eda090e770c/root,upperdir=/var/lib/docker/overlay/2a0f01d5a7455488043920ae09958a59cfc2cb188598e794b79452dee9207f19/upper,workdir=/var/lib/docker/overlay/2a0f01d5a7455488043920ae09958a59cfc2cb188598e794b79452dee9207f19/work)
overlay on /hostfs/var/lib/docker/overlay/f329e24fff6b080d6810ddd94fc48b90443c30db8568d3bf3c5aedc17ae83d88/merged type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay/dc59567af60f8c70b35c02692e2253904b22c6c9986b3df6fcf95eda090e770c/root,upperdir=/var/lib/docker/overlay/f329e24fff6b080d6810ddd94fc48b90443c30db8568d3bf3c5aedc17ae83d88/upper,workdir=/var/lib/docker/overlay/f329e24fff6b080d6810ddd94fc48b90443c30db8568d3bf3c5aedc17ae83d88/work)
overlay on /hostfs/var/lib/docker/overlay/f329e24fff6b080d6810ddd94fc48b90443c30db8568d3bf3c5aedc17ae83d88/merged type overlay (rw,relatime,lowerdir=/var/lib/docker/overlay/dc59567af60f8c70b35c02692e2253904b22c6c9986b3df6fcf95eda090e770c/root,upperdir=/var/lib/docker/overlay/f329e24fff6b080d6810ddd94fc48b90443c30db8568d3bf3c5aedc17ae83d88/upper,workdir=/var/lib/docker/overlay/f329e24fff6b080d6810ddd94fc48b90443c30db8568d3bf3c5aedc17ae83d88/work)
```

This would indicate that when docker launches the container, it's explicitly bind-mouting all the volumes within `/` as well. In which case yes, they can't be unmounted until the bind mounts are gone. This would indicate it's a docker problem, not telegraf's doing.

**Comment by m4ce** [comment]
@phemmer, yes.. it is a docker probem. The only reason why I opened this here is that if you want to monitor your host's volumes via telegraf running in docker, you must bind-mount / into the telegraf container as described in https://github.com/influxdata/telegraf/blob/master/plugins/inputs/system/DISK_README.md.

So, if people end up having the same issue, so far the only workaround is what I came up with in the first post.
